pkg/controller/jobframework/reconciler.go

```diff
@@ -412,15 +405,18 @@ func (r *JobReconciler) stopJob(ctx context.Context, job GenericJob, object clie
 		}
 	}
 
-	log.V(3).Info("restore podSetsInfo from annotation")
-	info, err := getPodSetsInfoFromObjectAnnotation(object, job)
-	if err != nil {
-		log.V(3).Error(err, "Unable to get original podSetsInfo")
-	} else {
+	if wl != nil {
+		log.V(3).Info("Restored job spec")
+		info := getPodSetsInfoFromWorkload(wl)
+		oldPS := job.PodSets()
 		job.RestorePodSetsInfo(info)
-		return r.client.Update(ctx, object)
+		if !equality.Semantic.DeepEqual(oldPS, job.PodSets) {
+			err := r.client.Update(ctx, object)
```

Comment:
why? In regular operations, we can assume that the Workload object is still there.

Reply:
With finalizers, which are cleared at the end of the 4step stop, if the job is suspended (step1 ok) and the finalizes are still present we know that at least the last one of the remaining steps  failed.

Reply:
Sure, that is one way to do it.

A better alternative is to call `stopJob` even if Job has `suspend=true`. Then internally `stopJob` makes all the steps that it could have missed.

Reply:
in which case , every time the the reconciler runs and the workload is not admitted?  We will end up doing two useless API calls every reconcile if we don't change the GenricJob interface to return a bool if a change needs to be made.

Reply:
We shouldn't issue the calls that are not needed (like if the objects wouldn't change), but we can still run through all the checks.

Reply:
In any case, we can do this in a follow up PR.